src/pybind/mgr/prometheus/module.py

```diff
@@ -76,7 +76,16 @@ def str_expfmt(self):
 
         def promethize(path):
             ''' replace illegal metric name characters '''
-            return path.replace('.', '_').replace('-', '_')
+            result = path.replace('.', '_').replace('+', '_plus') 
+
+            # Hyphens usually turn into underscores, unless they are
+            # trailing
+            if result.endswith("-"):
+                result = result[0:-1] + "_minus"
```

Comment:
what if there are some hyphens in the middle of the `result` *and* it is ending with another hyphen? probably we should put something like
```py
if result.endswith('-'):
  result = result.replace('-', '_')[0:-1] + '_minus'
else:
  result = result.replace('-', '_')
```

Reply:
Good point, @tchaikov — would the case still hold for '+'?

I would think what we'd want is there to be no namespace collisions due to replacing (n) special character types with (n-1) allowed characters, but there's a limited subset available. Is this a check that is worth doing?

Reply:
i am not sure. per the discussion in http://tracker.ceph.com/issues/20899, what we need is to encode the pathnames so that they don't contain "+" or "-". and your patch also replaces "." with "_". 

@jcsp makes the ending "-" as an exception, by replacing it with "_minus" instead of "_". i made last comment according to the code comment. actually, i have no idea why we need to do so.

Reply:
The "." was the previous code before my patch. The problem with simply replacing every illegal character with _ is that you run into namespace collisions… see https://github.com/ceph/ceph/pull/17318/commits/d719cd04b294e90ab9d440ba7d033826c069a2de

So there seems to be a good reason to spell out - and + when they're final, as this accommodates the existing mds metrics.

Reply:
@jermudgeon thanks for explaining this in details! i get it now. so 

> So there seems to be a good reason to spell out - and + when they're final, as this accommodates the existing mds metrics.

makes sense to me. but my only concern is that, in theory, "+" is not allowed in the prometheus metrics names (or whatever it is called). so, *what if*, ceph emits metrics with "+" in the middle of their name?

so i think, it seems that the code in this PR replaces all the occurrence of "+" with "\_plus" unnecessarily even we could only check the ending "+". this change does cover most of the cases we can suffer from the "+", "\-", and "\."